core/src/main/java/org/elasticsearch/search/profile/InternalProfileResult.java

```diff
@@ -0,0 +1,193 @@
+ */
+final class InternalProfileResult implements ProfileResult, Streamable, ToXContent {
+
+    private static final ParseField QUERY_TYPE = new ParseField("query_type");
+    private static final ParseField LUCENE_DESCRIPTION = new ParseField("lucene");
+    private static final ParseField NODE_TIME = new ParseField("time");
+    private static final ParseField CHILDREN = new ParseField("children");
+    private static final ParseField BREAKDOWN = new ParseField("breakdown");
+
+    private String queryType;
+    private String luceneDescription;
+    private Map<String, Long> timings;
+    private long nodeTime = -1;     // Use -1 instead of Null so it can be serialized, and there should never be a negative time
+    private ArrayList<InternalProfileResult> children;
+
+    public InternalProfileResult(Query query, Map<String, Long> timings) {
+        children = new ArrayList<>(5);
+        this.queryType = query.getClass().getSimpleName();
+        this.luceneDescription = query.toString();
+        this.timings = timings;
+    }
+
+    public InternalProfileResult() {
+
+    }
+
+    /**
+     * Add a child profile result to this node
+     * @param child The child to add
+     */
+    public void addChild(InternalProfileResult child) {
+        children.add(child);
+    }
+
+    /**
+     * Retrieve a list of all profiled children at this node
+     * @return List of profiled children
+     */
+    public ArrayList<InternalProfileResult> getChildren() {
+        return children;
+    }
+
+    @Override
+    public String getLuceneDescription() {
+        return luceneDescription;
+    }
+
+    @Override
+    public String getQueryName() {
+        return queryType;
+    }
+
+    @Override
+    public Map<String, Long> getTimeBreakdown() {
+        return timings;
+    }
+
+    @Override
+    public long getTime() {
+        if (nodeTime != -1) {
+            return nodeTime;
+        }
+
+        nodeTime = 0;
+        for (long time : timings.values()) {
+            nodeTime += time;
+        }
+
+        // Then add up our children
+        for (InternalProfileResult child : children) {
+            child.getTime();
+        }
+
+        return nodeTime;
+    }
+
+    /**
+     * Static helper to read an InternalProfileResult off the stream
+     */
+    public static InternalProfileResult readProfileResult(StreamInput in) throws IOException {
+        InternalProfileResult newResults = new InternalProfileResult();
+        newResults.readFrom(in);
+        return newResults;
+    }
+
+    @Override
+    public List<ProfileResult> getProfiledChildren() {
+        return Collections.unmodifiableList(children);
+    }
+
+    @Override
+    public void readFrom(StreamInput in) throws IOException {
+        queryType = in.readString();
+        luceneDescription = in.readString();
+        nodeTime = in.readLong();
+
+        int timingsSize = in.readVInt();
+        timings = new HashMap<>(timingsSize);
```

Comment:
Hrm yeah.  I'm not sure how to prevent this?  Are there examples elsewhere of serializing an unordered map?  Dump to a list and sort first or something?

I'm not sure if would actually matter, since none of the operations care about the order of ProfileResults.  As long as each individual one deserializes correctly I *think* it would be ok.  I'll see if I can write up a test to see what happens...